File: src/typing/func_stmt_config.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

module Ast = Flow_ast
module Flow = Flow_js
open Reason

module Make (Destructuring : Destructuring_sig.S) (Statement : Statement_sig.S) :
  Func_stmt_config_sig.S with module Types = Func_stmt_config_types.Types = struct
  module Types = Func_stmt_config_types.Types
  open Types

  let param_type (Param { t; pattern; default; _ }) =
    match pattern with
    | Id id ->
      let { Ast.Pattern.Identifier.name = (_, { Ast.Identifier.name; _ }); optional; _ } = id in
      let t =
        if optional || default <> None then
          TypeUtil.optional t
        else
          t
      in
      (Some name, t)
    | _ ->
      let t =
        if default <> None then
          TypeUtil.optional t
        else
          t
      in
      (None, t)

  let rest_type (Rest { t; loc; id; _ }) =
    let { Ast.Pattern.Identifier.name = (_, { Ast.Identifier.name; _ }); _ } = id in
    (Some name, loc, t)

  let this_type (This { t; _ }) = t

  let is_param_type_annotated (Param { has_anno; _ }) = has_anno

  let is_rest_type_annotated (Rest { has_anno; _ }) = has_anno

  let subst_param cx map param =
    let (Param { t; loc; ploc; pattern; default; has_anno }) = param in
    let t = Flow.subst cx map t in
    Param { t; loc; ploc; pattern; default; has_anno }

  let subst_rest cx map rest =
    let (Rest { t; loc; ploc; id; has_anno }) = rest in
    let t = Flow.subst cx map t in
    Rest { t; loc; ploc; id; has_anno }

  let subst_this cx map (This { t; loc; annot }) =
    let t = Flow.subst cx map t in
    This { t; loc; annot }

  let destruct cx ~use_op ~name_loc name default t =
    Base.Option.iter
      ~f:(fun d ->
        let reason = mk_reason (RIdentifier (OrdinaryName name)) name_loc in
        let default_t = Flow.mk_default cx reason d in
        Flow.flow cx (default_t, Type.UseT (use_op, t)))
      default;
    t

  let flow_default cx annot_t default_t =
    let open Type in
    let use_op =
      Op
        (AssignVar
           { var = Some (TypeUtil.reason_of_t annot_t); init = TypeUtil.reason_of_t default_t }
        )
    in
    Flow.flow cx (default_t, UseT (use_op, annot_t))

  let eval_default cx ?(always_flow_default = false) annot_t has_anno =
    Base.Option.map ~f:(fun e ->
        let (((loc, default_t), expr) as e) = Statement.expression cx e in
        if has_anno then (
          flow_default cx annot_t default_t;
          ((loc, annot_t), expr)
        ) else (
          if always_flow_default then flow_default cx annot_t default_t;
          e
        )
    )

  let eval_param cx (Param { t; loc; ploc; pattern; default; has_anno }) =
    match pattern with
    | Id ({ Ast.Pattern.Identifier.name = ((name_loc, _), { Ast.Identifier.name; _ }); _ } as id) ->
      let reason = mk_reason (RIdentifier (OrdinaryName name)) name_loc in
      let t = Env.find_write cx Env_api.OrdinaryNameLoc reason in
      let default = eval_default cx ~always_flow_default:true t has_anno default in
      (loc, { Ast.Function.Param.argument = ((ploc, t), Ast.Pattern.Identifier id); default })
    | Object { annot; properties; comments } ->
      let default = eval_default cx t has_anno default in
      let properties =
        let init = Destructuring.empty t ~annot:has_anno in
        let f = destruct cx in
        Destructuring.object_properties cx ~f ~parent_loc:ploc init properties
      in
      ( loc,
        {
          Ast.Function.Param.argument =
            ((ploc, t), Ast.Pattern.Object { Ast.Pattern.Object.properties; annot; comments });
          default;
        }
      )
    | Array { annot; elements; comments } ->
      let default = eval_default cx t has_anno default in
      let elements =
        let init = Destructuring.empty t ~annot:has_anno in
        let f = destruct cx in
        Destructuring.array_elements cx ~f init elements
      in
      ( loc,
        {
          Ast.Function.Param.argument =
            ((ploc, t), Ast.Pattern.Array { Ast.Pattern.Array.elements; annot; comments });
          default;
        }
      )

  let eval_rest cx (Rest { t = _; loc; ploc; id; has_anno = _ }) =
    let { Ast.Pattern.Identifier.name = ((name_loc, _), { Ast.Identifier.name; _ }); _ } = id in
    let reason = mk_reason (RIdentifier (OrdinaryName name)) name_loc in
    let t = Env.find_write cx Env_api.OrdinaryNameLoc reason in
    ( loc,
      { Ast.Function.RestParam.argument = ((ploc, t), Ast.Pattern.Identifier id); comments = None }
    )

  let eval_this _cx (This { t = _; annot; loc }) =
    (loc, { Ast.Function.ThisParam.annot; comments = None })
end

File: rust_port/crates/flow_typing_statement/src/func_stmt_config.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use dupe::Dupe;
use flow_aloc::ALoc;
use flow_common::reason;
use flow_common::reason::Name;
use flow_common::reason::VirtualReasonDesc;
use flow_common::subst_name::SubstName;
use flow_data_structure_wrapper::ord_map::FlowOrdMap;
use flow_env_builder::env_api;
use flow_parser::ast;
use flow_parser::ast::function;
use flow_parser::ast::pattern;
use flow_typing_context::Context;
use flow_typing_flow_js::flow_js;
use flow_typing_loc_env::func_stmt_config_types::Param;
use flow_typing_loc_env::func_stmt_config_types::Pattern;
use flow_typing_loc_env::func_stmt_config_types::Rest;
use flow_typing_loc_env::func_stmt_config_types::ThisParam;
use flow_typing_type::type_;
use flow_typing_type::type_::FunParam;
use flow_typing_type::type_::FunRestParam;
use flow_typing_type::type_::Type;
use flow_typing_type::type_::UseOp;
use flow_typing_type::type_util;
use flow_typing_utils::abnormal::CheckExprError;
use flow_typing_utils::type_env;

use crate::destructuring;
use crate::statement;

pub fn param_type(param: &Param) -> FunParam {
    match &param.pattern {
        Pattern::Id(id) => {
            let name = &id.name.name;
            let optional = id.optional;
            let t = if optional || param.default.is_some() {
                type_util::optional(param.t.dupe(), None, false)
            } else {
                param.t.dupe()
            };
            FunParam(Some(name.dupe()), t)
        }
        Pattern::Object { optional, .. } | Pattern::Array { optional, .. } => {
            let t = if *optional || param.default.is_some() {
                type_util::optional(param.t.dupe(), None, false)
            } else {
                param.t.dupe()
            };
            FunParam(None, t)
        }
        Pattern::ParamPropertyPattern(_) => {
            let t = if param.default.is_some() {
                type_util::optional(param.t.dupe(), None, false)
            } else {
                param.t.dupe()
            };
            FunParam(None, t)
        }
    }
}

pub fn rest_type(rest: &Rest) -> FunRestParam {
    let name = &rest.id.name.name;
    FunRestParam(Some(name.dupe()), rest.loc.dupe(), rest.t.dupe())
}

pub fn this_type(this: &ThisParam) -> Type {
    this.t.dupe()
}

pub fn is_param_type_annotated(param: &Param) -> bool {
    param.has_anno
}

pub fn is_rest_type_annotated(rest: &Rest) -> bool {
    rest.has_anno
}

pub fn subst_param<'a>(
    cx: &Context<'a>,
    map: &FlowOrdMap<SubstName, Type>,
    param: &Param,
) -> Param {
    let t = flow_js::subst(cx, None, None, None, map, param.t.dupe());
    Param {
        t,
        loc: param.loc.dupe(),
        ploc: param.ploc.dupe(),
        pattern: param.pattern.clone(),
        default: param.default.dupe(),
        has_anno: param.has_anno,
    }
}

pub fn subst_rest<'a>(cx: &Context<'a>, map: &FlowOrdMap<SubstName, Type>, rest: &Rest) -> Rest {
    let t = flow_js::subst(cx, None, None, None, map, rest.t.dupe());
    Rest {
        t,
        loc: rest.loc.dupe(),
        ploc: rest.ploc.dupe(),
        id: rest.id.clone(),
        has_anno: rest.has_anno,
    }
}

pub fn subst_this<'a>(
    cx: &Context<'a>,
    map: &FlowOrdMap<SubstName, Type>,
    this: &ThisParam,
) -> ThisParam {
    let t = flow_js::subst(cx, None, None, None, map, this.t.dupe());
    ThisParam {
        t,
        loc: this.loc.dupe(),
        annot: this.annot.clone(),
    }
}

fn destruct<'a>(
    cx: &Context<'a>,
    use_op: &UseOp,
    name_loc: ALoc,
    name: &str,
    default: Option<&flow_typing_default::Default<Type>>,
    t: Type,
) -> Result<Type, flow_utils_concurrency::job_error::JobError> {
    if let Some(d) = default {
        let reason = reason::mk_reason(VirtualReasonDesc::RIdentifier(Name::new(name)), name_loc);
        let default_t = flow_js::mk_default_non_speculating(cx, &reason, d)?;
        flow_js::flow_non_speculating(
            cx,
            (
                &default_t,
                &type_::UseT::new(type_::UseTInner::UseT(use_op.dupe(), t.dupe())),
            ),
        )?;
    }
    Ok(t)
}

fn flow_default<'a>(
    cx: &Context<'a>,
    annot_t: &Type,
    default_t: &Type,
) -> Result<(), flow_utils_concurrency::job_error::JobError> {
    let use_op = UseOp::Op(std::sync::Arc::new(type_::RootUseOp::AssignVar {
        var: Some(type_util::reason_of_t(annot_t).dupe()),
        init: type_util::reason_of_t(default_t).dupe(),
    }));
    flow_js::flow_non_speculating(
        cx,
        (
            default_t,
            &type_::UseT::new(type_::UseTInner::UseT(use_op, annot_t.dupe())),
        ),
    )?;
    Ok(())
}

fn eval_default<'a>(
    cx: &Context<'a>,
    always_flow_default: bool,
    annot_t: &Type,
    has_anno: bool,
    default: Option<ast::expression::Expression<ALoc, ALoc>>,
) -> Result<Option<ast::expression::Expression<ALoc, (ALoc, Type)>>, CheckExprError> {
    default
        .map(|e| {
            let e = statement::expression(None, None, None, cx, &e)?;
            let (loc, default_t) = e.loc().dupe();
            if has_anno {
                flow_default(cx, annot_t, &default_t)?;
                let mut inner = (*e.0).clone();
                *inner.loc_mut() = (loc, annot_t.dupe());
                Ok(ast::expression::Expression(inner.into()))
            } else {
                if always_flow_default {
                    flow_default(cx, annot_t, &default_t)?;
                }
                Ok(e)
            }
        })
        .transpose()
}

pub fn eval_param<'a>(
    cx: &Context<'a>,
    param: &Param,
) -> Result<flow_typing_loc_env::func_stmt_config_types::ParamAst<(ALoc, Type)>, CheckExprError> {
    let Param {
        t,
        loc,
        ploc,
        pattern,
        default,
        has_anno,
    } = param;
    match pattern {
        Pattern::Id(id) => {
            let name_loc = id.name.loc.0.dupe();
            let name = &id.name.name;
            let reason = reason::mk_reason(
                VirtualReasonDesc::RIdentifier(Name::new(name.dupe())),
                name_loc,
            );
            let t = type_env::find_write(cx, env_api::DefLocType::OrdinaryNameLoc, reason);
            let default = eval_default(cx, true, &t, *has_anno, default.dupe())?;
            Ok(function::Param::RegularParam {
                loc: loc.dupe(),
                argument: pattern::Pattern::Identifier {
                    loc: (ploc.dupe(), t),
                    inner: id.clone().into(),
                },
                default,
            })
        }
        Pattern::Object {
            annot,
            properties,
            optional,
            comments,
        } => {
            let default = eval_default(cx, false, t, *has_anno, default.dupe())?;
            let mut init = destructuring::empty(None, None);
            let properties = destructuring::object_properties(
                cx,
                &|use_op, name_loc, name, default, t| {
                    destruct(cx, use_op, name_loc, name, default, t)
                },
                ploc.dupe(),
                &mut init,
                properties,
            )?;
            Ok(function::Param::RegularParam {
                loc: loc.dupe(),
                argument: pattern::Pattern::Object {
                    loc: (ploc.dupe(), t.dupe()),
                    inner: pattern::Object {
                        properties: properties.into(),
                        annot: annot.clone(),
                        optional: *optional,
                        comments: comments.dupe(),
                    }
                    .into(),
                },
                default,
            })
        }
        Pattern::Array {
            annot,
            elements,
            optional,
            comments,
        } => {
            let default = eval_default(cx, false, t, *has_anno, default.dupe())?;
            let mut init = destructuring::empty(None, None);
            let elements = destructuring::array_elements(
                cx,
                &|use_op, name_loc, name, default, t| {
                    destruct(cx, use_op, name_loc, name, default, t)
                },
                &mut init,
                elements,
            )?;
            Ok(function::Param::RegularParam {
                loc: loc.dupe(),
                argument: pattern::Pattern::Array {
                    loc: (ploc.dupe(), t.dupe()),
                    inner: pattern::Array {
                        elements: elements.into(),
                        annot: annot.clone(),
                        optional: *optional,
                        comments: comments.dupe(),
                    }
                    .into(),
                },
                default,
            })
        }
        Pattern::ParamPropertyPattern(prop) => Ok(function::Param::ParamProperty {
            loc: loc.dupe(),
            property: prop.clone(),
        }),
    }
}

pub fn eval_rest<'a>(
    cx: &Context<'a>,
    rest: &Rest,
) -> flow_typing_loc_env::func_stmt_config_types::RestAst<(ALoc, Type)> {
    let Rest {
        t: _,
        loc,
        ploc,
        id,
        has_anno: _,
    } = rest;
    let name_loc = id.name.loc.0.dupe();
    let name = &id.name.name;
    let reason = reason::mk_reason(
        VirtualReasonDesc::RIdentifier(Name::new(name.dupe())),
        name_loc,
    );
    let t = type_env::find_write(cx, env_api::DefLocType::OrdinaryNameLoc, reason);
    function::RestParam {
        loc: loc.dupe(),
        argument: pattern::Pattern::Identifier {
            loc: (ploc.dupe(), t),
            inner: id.clone().into(),
        },
        comments: None,
    }
}

pub fn eval_this<'a>(
    _cx: &Context<'a>,
    this: &ThisParam,
) -> flow_typing_loc_env::func_stmt_config_types::ThisAst<(ALoc, Type)> {
    ast::function::ThisParam {
        loc: this.loc.dupe(),
        annot: this.annot.clone(),
        comments: None,
    }
}
